app/services/cart.js

```diff
@@ -1,179 +1,382 @@
-      }
+  /**
+   * When the user logs in we
+   *    - fetch the user's cart items
+   *    - push any record he/she might have added as a guest before
+   *   logging in
+   */
+  onLogin() {
+    this.populate().then(() => {
+      this.persistLocalRecords();
+    });
+  },
+
+  /**
+   * When logging out, we clear the local data
+   */
+  onLogout() {
+    this.get("store")
+      .peekAll("cart_item")
+      .forEach(record => this.get("store").unloadRecord(record));
+  },
+
+  // ---- Control methods
+
+  /**
+   * Checkout the cart and fill the order
+   */
+  checkoutOrder(order) {
+    return this.POST(`/cart_items/checkout`, {
+      order_id: order.get("id")
+    }).then(data => {
+      this.get("store").pushPayload(data);
+      return this.refresh();
+    });
+  },
+
+  /**
+   * Fetches the user's cart items from the api
+   */
+  populate() {
+    if (!this.get("isLoggedIn")) {
+      return Ember.RSVP.resolve();
     }
+    return this.get("store").findAll("cart_item", { reload: true });
   },
 
-  setItemAvailability(cartItem, available) {
-    Ember.set(cartItem, "available", available);
-    this.pushItem(cartItem);
+  /**
+   * Clears the local data and re-fetches the data
+   */
+  refresh() {
+    this.get("store").unloadAll("cart_item");
+    return this.populate();
   },
 
-  content: computed(function() {
-    return JSON.parse(window.localStorage.getItem("cart") || "[]");
-  }),
+  /**
+   * Cart items that were added as guest are pushed to the db
+   */
+  persistLocalRecords() {
+    if (!this.get("isLoggedIn")) {
+      return Ember.RSVP.resolve();
+    }
 
-  cartItems: computed("[]", "@each.available", function() {
-    var content = this.get("content");
-    var allItems = [];
-    content.forEach(record => {
-      if (record.available) {
-        allItems.push(
-          this.get("store").peekRecord(record.modelType, record.id)
-        );
-      }
+    this.clearDuplicates();
+
+    return Ember.RSVP.all(
+      this.get("offlineCartItems").map(it => {
+        it.set("user", this.get("user"));
+        it.set("userId", this.get("userId"));
+        return it.save();
+      })
+    ).then(res => {
+      this.forgetGuestItems();
+      return res;
     });
-    return allItems;
-  }),
+  },
 
-  packageIds: computed("cartItems", function() {
-    let package_ids = [];
-    this.get("cartItems").forEach(record => {
-      if (record) {
-        let ids = record.get("isItem")
-          ? record.get("packages").map(pkg => pkg.get("id"))
-          : [record.get("id")];
-        package_ids = package_ids.concat(ids);
-      }
+  add(pkgOrItem) {
+    if (pkgOrItem.get("isItem")) {
+      return this.addItem(pkgOrItem);
+    }
+    return this.addPackage(pkgOrItem);
+  },
+
+  /**
+   * Adds the packages of the item to the cloud cart.
+   * If the user is not logged in, will only save it locally
+   */
+  addItem(item) {
+    return Ember.RSVP.all(
+      item.get("packages").map(pkg => this.addPackage(pkg))
```

Comment:
is it fine to use `all` here ? I think it should be used for multiple promises here it is one promise having multiple promises. Rectify me if I am wrong.

Reply:
`map` returns an array, so it is multiple promises. one for each package :)